File: backend/engine/scoring_engine.py
"""
GridForge Scoring Engine — Python port of src/engine/ScoringEngine.js

Pure-function scoring engine for role-specific + system-wide player evaluation.
"""

from __future__ import annotations
import math

from .constants import SCORING_CONFIG
from .utils import clamp


# ─── Piecewise Linear Interpolation ───

def map_threshold(value: float, breakpoints: list[list[float]]) -> float:
    if not breakpoints:
        return 50.0
    if len(breakpoints) == 1:
        return breakpoints[0][1]

    if value <= breakpoints[0][0]:
        return breakpoints[0][1]
    if value >= breakpoints[-1][0]:
        return breakpoints[-1][1]

    for i in range(len(breakpoints) - 1):
        x0, y0 = breakpoints[i]
        x1, y1 = breakpoints[i + 1]
        if x0 <= value <= x1:
            if x1 == x0:
                return y0
            t = (value - x0) / (x1 - x0)
            return y0 + t * (y1 - y0)

    return breakpoints[-1][1]


# ─── ROLE SCORE COMPUTATION ───

def compute_role_score(role: str, stats: dict) -> dict:
    cfg = SCORING_CONFIG.get(role, SCORING_CONFIG["GENERATOR"])

    dispatch = {
        "TRADER": _trader_role_score,
        "GENERATOR": _generator_role_score,
        "BESS": _bess_role_score,
        "SUPPLIER": _supplier_role_score,
        "DSR": _dsr_role_score,
        "NESO": lambda s, c: _neso_role_score(s, SCORING_CONFIG["NESO"]),
        "ELEXON": lambda s, c: _elexon_role_score(s, SCORING_CONFIG["ELEXON"]),
        "INTERCONNECTOR": _interconnector_role_score,
    }

    fn = dispatch.get(role, lambda s, c: _generator_role_score(s, SCORING_CONFIG["GENERATOR"]))
    return fn(stats, cfg)


# ── Trader ──
def _trader_role_score(stats: dict, cfg: dict) -> dict:
    rar = stats.get("netProfit", 0) / max(1, stats.get("maxDrawdown", 1))
    primary_score = clamp(map_threshold(rar, cfg.get("breakpoints", [])), 0, 100)

    margin_evt_penalty = min(stats.get("marginEvents", 0), 10)
    secondary_score = clamp(100 - cfg.get("marginPenalty", 10) * margin_evt_penalty, 0, 100)

    pw = cfg.get("primaryWeight", 0.85)
    role_score = clamp(round(pw * primary_score + (1 - pw) * secondary_score), 0, 100)

    return {
        "roleScore": role_score,
        "primary": {"name": "Risk-Adjusted Return", "value": round(rar, 2), "score": round(primary_score)},
        "secondary": [
            {"name": "Margin Events", "value": stats.get("marginEvents", 0), "score": round(secondary_score)},
        ],
    }


# ── Generator ──
def _generator_role_score(stats: dict, cfg: dict) -> dict:
    cap_mw = stats.get("capacityMW", 1)
    profit_per_mw = stats.get("netProfit", 0) / cap_mw if cap_mw > 0 else 0
    primary_score = clamp(map_threshold(profit_per_mw, cfg.get("breakpoints", [])), 0, 100)

    imb_ref = 50
    total_mwh = stats.get("totalMWh", 1)
    imb_cost_per_mwh = abs(stats.get("imbalanceCost", 0)) / total_mwh if total_mwh > 0 else 0
    secondary_score = clamp(100 - (imb_cost_per_mwh / imb_ref) * 100, 0, 100)

    pw = cfg.get("primaryWeight", 0.80)
    role_score = clamp(round(pw * primary_score + (1 - pw) * secondary_score), 0, 100)

    return {
        "roleScore": role_score,
        "primary": {"name": "Profit/MW", "value": round(profit_per_mw), "score": round(primary_score)},
        "secondary": [
            {"name": "Imbalance Cost/MWh", "value": round(imb_cost_per_mwh, 1), "score": round(secondary_score)},
        ],
    }


# ── BESS ──
def _bess_role_score(stats: dict, cfg: dict) -> dict:
    mwh_shifted = max(1, stats.get("mwhShifted", 1))
    rev_per_mwh = stats.get("netProfit", 0) / mwh_shifted
    primary_score = clamp(map_threshold(rev_per_mwh, cfg.get("breakpoints", [])), 0, 100)

    total_revenue = stats.get("totalRevenue", 1)
    bm_share = stats.get("bmRevenue", 0) / total_revenue if total_revenue > 0 else 0
    bm_share_score = clamp(bm_share * 200, 0, 100)

    soc_penalties = stats.get("socPenalties", 0)
    soc_score = clamp(100 - soc_penalties * 20, 0, 100)

    secondary_score = (bm_share_score + soc_score) / 2

    pw = cfg.get("primaryWeight", 0.75)
    role_score = clamp(round(pw * primary_score + (1 - pw) * secondary_score), 0, 100)

    return {
        "roleScore": role_score,
        "primary": {"name": "£/MWh Shifted", "value": round(rev_per_mwh, 1), "score": round(primary_score)},
        "secondary": [
            {"name": "BM Revenue Share", "value": f"{round(bm_share * 100)}%", "score": round(bm_share_score)},
            {"name": "SoC Health", "value": f"{soc_penalties} penalties", "score": round(soc_score)},
        ],
    }


# ── Supplier ──
def _supplier_role_score(stats: dict, cfg: dict) -> dict:
    total_mwh = stats.get("totalMWh", 1)
    cost_per_mwh = abs(stats.get("netCost", 0)) / total_mwh if total_mwh > 0 else 80
    primary_score = clamp(map_threshold(cost_per_mwh, cfg.get("breakpoints", [])), 0, 100)

    hedge_ratio = clamp(stats.get("hedgeRatio", 0.5) * 100, 0, 100)
    net_cost = stats.get("netCost", 1)
    imb_pct = abs(stats.get("imbalanceCost", 0)) / net_cost * 100 if net_cost > 0 else 50
    imb_pct_score = clamp(100 - imb_pct * 2, 0, 100)

    secondary_score = (hedge_ratio + imb_pct_score) / 2

    pw = cfg.get("primaryWeight", 0.80)
    role_score = clamp(round(pw * primary_score + (1 - pw) * secondary_score), 0, 100)

    return {
        "roleScore": role_score,
        "primary": {"name": "Cost/MWh", "value": f"£{round(cost_per_mwh)}", "score": round(primary_score)},
        "secondary": [
            {"name": "Hedge Ratio", "value": f"{round(stats.get('hedgeRatio', 0) * 100)}%", "score": round(hedge_ratio)},
            {"name": "Imbalance % of Cost", "value": f"{round(imb_pct)}%", "score": round(imb_pct_score)},
        ],
    }


# ── DSR ──
def _dsr_role_score(stats: dict, cfg: dict) -> dict:
    reliability = clamp(stats.get("reliability", 1), 0, 1)
    rel_adj_rev = stats.get("netProfit", 0) * reliability
    primary_score = clamp(map_threshold(rel_adj_rev, cfg.get("breakpoints", [])), 0, 100)

    missed_events = stats.get("missedEvents", 0)
    secondary_score = clamp(100 - missed_events * 15, 0, 100)

    pw = cfg.get("primaryWeight", 0.80)
    role_score = clamp(round(pw * primary_score + (1 - pw) * secondary_score), 0, 100)

    return {
        "roleScore": role_score,
        "primary": {"name": "Reliability-Adj Rev", "value": f"£{round(rel_adj_rev)}", "score": round(primary_score)},
        "secondary": [
            {"name": "Missed Events", "value": missed_events, "score": round(secondary_score)},
        ],
    }


# ── NESO ──
def _neso_role_score(stats: dict, cfg: dict) -> dict:
    avg_abs_niv = stats.get("avgAbsNIV", 0)
    stability_score = clamp(100 - (avg_abs_niv / 6.2), 0, 100)

    total_sps = max(1, stats.get("totalSPs", 1))
    cost_per_sp = stats.get("totalSystemCost", 0) / total_sps
    cost_score = clamp(100 - (cost_per_sp / 50), 0, 100)

    mae_score = clamp(100 - (stats.get("forecastMAE", 0)) * 2, 0, 100)

    price_vol = stats.get("priceVolatility", 0)
    price_vol_score = clamp(100 - price_vol * 0.5, 0, 100)
    participation = clamp(stats.get("participationRate", 0.5) * 100, 0, 100)
    clearing_score = (price_vol_score + participation) / 2

    sw = cfg.get("stabilityWeight", 0.40)
    cw = cfg.get("costWeight", 0.20)
    mw = cfg.get("maeWeight", 0.15)
    clw = cfg.get("clearingWeight", 0.25)
    role_score = clamp(round(sw * stability_score + cw * cost_score + mw * mae_score + clw * clearing_score), 0, 100)

    return {
        "roleScore": role_score,
        "primary": {"name": "Stability Index", "value": round(stability_score), "score": round(stability_score)},
        "secondary": [
            {"name": "System Cost/SP", "value": f"£{round(cost_per_sp)}", "score": round(cost_score)},
            {"name": "Forecast MAE", "value": round(stats.get("forecastMAE", 0), 1), "score": round(mae_score)},
            {"name": "Clearing Quality", "value": round(clearing_score), "score": round(clearing_score)},
        ],
    }


# ── Elexon ──
def _elexon_role_score(stats: dict, cfg: dict) -> dict:
    settlement_error = stats.get("settlementError", 0)
    accuracy_score = clamp(100 - settlement_error * 5, 0, 100)

    on_time_rate = clamp(stats.get("onTimeRate", 1.0) * 100, 0, 100)
    audit_coverage = clamp(stats.get("auditCoverage", 1.0) * 100, 0, 100)

    aw = cfg.get("accuracyWeight", 0.50)
    tw = cfg.get("timelinessWeight", 0.30)
    trw = cfg.get("transparencyWeight", 0.20)
    role_score = clamp(round(aw * accuracy_score + tw * on_time_rate + trw * audit_coverage), 0, 100)

    return {
        "roleScore": role_score,
        "primary": {"name": "Settlement Accuracy", "value": f"{round(accuracy_score)}%", "score": round(accuracy_score)},
        "secondary": [
            {"name": "On-Time Rate", "value": f"{round(on_time_rate)}%", "score": round(on_time_rate)},
            {"name": "Audit Coverage", "value": f"{round(audit_coverage)}%", "score": round(audit_coverage)},
        ],
    }


# ── Interconnector ──
def _interconnector_role_score(stats: dict, cfg: dict) -> dict:
    congestion_rev = stats.get("congestionRevenue", 0)
    primary_score = clamp(map_threshold(congestion_rev, cfg.get("breakpoints", [])), 0, 100)

    availability = clamp(stats.get("availability", 1) * 100, 0, 100)
    stress_help = clamp(stats.get("stressContribution", 0) * 20, 0, 100)
    secondary_score = (availability + stress_help) / 2

    pw = cfg.get("primaryWeight", 0.80)
    role_score = clamp(round(pw * primary_score + (1 - pw) * secondary_score), 0, 100)

    return {
        "roleScore": role_score,
        "primary": {"name": "Congestion Revenue", "value": f"£{round(congestion_rev)}", "score": round(primary_score)},
        "secondary": [
            {"name": "Availability", "value": f"{round(availability)}%", "score": round(availability)},
            {"name": "Stress Help", "value": round(stats.get("stressContribution", 0), 1), "score": round(stress_help)},
        ],
    }


# ─── SYSTEM SCORE ───

def compute_system_score(metrics: dict | None) -> int:
    if not metrics:
        return 50

    niv_base = clamp(50 + (metrics.get("totalNIVContribution", 0)) * 0.1, 0, 100)
    stress_bonus = min((metrics.get("stressWindowHelps", 0)) * 5, 25)
    missed_penalty = (metrics.get("missedDeliveries", 0)) * 10
    blackout_penalty = 40 if metrics.get("causedBlackout") else 0

    return clamp(round(niv_base + stress_bonus - missed_penalty - blackout_penalty), 0, 100)


# ─── OVERALL SCORE ───

def compute_overall_score(role_score: int, system_score: int, alpha: float | None = None) -> int:
    a = alpha if alpha is not None else SCORING_CONFIG.get("alpha", 0.6)
    return clamp(round(a * role_score + (1 - a) * system_score), 0, 100)


# ─── MULTI-ROUND FINAL SCORE ───

def compute_final_score(overall_scores: list[float]) -> int:
    if not overall_scores:
        return 0
    n = len(overall_scores)
    mean = sum(overall_scores) / n
    if n < 2:
        return round(mean)

    variance = sum((v - mean) ** 2 for v in overall_scores) / (n - 1)
    std = math.sqrt(variance)
    penalty = SCORING_CONFIG.get("consistencyPenalty", 0.1)

    return clamp(round(mean - penalty * std), 0, 100)

File: frontend/src/engine/ScoringEngine.js
// ─── SCORING ENGINE ───
// Pure-function scoring engine for role-specific + system-wide player evaluation.
// No React dependencies — importable from anywhere.

import { SCORING_CONFIG } from '../shared/constants.js';

// ─── Piecewise Linear Interpolation ───
// breakpoints: [[x0,y0],[x1,y1],...] sorted ascending by x
// Returns interpolated y for given x, clamped to edge values.
export function mapThreshold(value, breakpoints) {
    if (!breakpoints || breakpoints.length === 0) return 50;
    if (breakpoints.length === 1) return breakpoints[0][1];

    // Below first breakpoint
    if (value <= breakpoints[0][0]) return breakpoints[0][1];
    // Above last breakpoint
    if (value >= breakpoints[breakpoints.length - 1][0]) return breakpoints[breakpoints.length - 1][1];

    // Find bounding pair and interpolate
    for (let i = 0; i < breakpoints.length - 1; i++) {
        const [x0, y0] = breakpoints[i];
        const [x1, y1] = breakpoints[i + 1];
        if (value >= x0 && value <= x1) {
            if (x1 === x0) return y0;
            const t = (value - x0) / (x1 - x0);
            return y0 + t * (y1 - y0);
        }
    }
    return breakpoints[breakpoints.length - 1][1];
}

// Clamp utility
const clamp = (v, lo, hi) => Math.max(lo, Math.min(hi, v));

// ─── ROLE SCORE COMPUTATION ───
// Each role has a primary KPI (80% weight) and supporting KPIs (20% weight).
// Returns { roleScore: 0-100, primary: {name, value, score}, secondary: [{name, value, score}] }

export function computeRoleScore(role, stats) {
    const cfg = SCORING_CONFIG[role] || SCORING_CONFIG.GENERATOR;

    switch (role) {
        case 'TRADER': return traderRoleScore(stats, cfg);
        case 'GENERATOR': return generatorRoleScore(stats, cfg);
        case 'BESS': return bessRoleScore(stats, cfg);
        case 'SUPPLIER': return supplierRoleScore(stats, cfg);
        case 'DSR': return dsrRoleScore(stats, cfg);
        case 'NESO': return nesoRoleScore(stats, SCORING_CONFIG.NESO);
        case 'ELEXON': return elexonRoleScore(stats, SCORING_CONFIG.ELEXON);
        case 'INTERCONNECTOR': return interconnectorRoleScore(stats, cfg);
        default: return generatorRoleScore(stats, SCORING_CONFIG.GENERATOR);
    }
}

// ── Trader ──
// Primary: Risk-Adjusted Return = NetProfit / max(1, MaxDrawdown)
// Secondary: Margin events penalty (-10 per event)
function traderRoleScore(stats, cfg) {
    const rar = stats.netProfit / Math.max(1, stats.maxDrawdown || 1);
    const primaryScore = clamp(mapThreshold(rar, cfg.breakpoints), 0, 100);

    const marginEvtPenalty = Math.min(stats.marginEvents || 0, 10);
    const secondaryScore = clamp(100 - (cfg.marginPenalty || 10) * marginEvtPenalty, 0, 100);

    const pw = cfg.primaryWeight || 0.85;
    const roleScore = clamp(Math.round(pw * primaryScore + (1 - pw) * secondaryScore), 0, 100);

    return {
        roleScore,
        primary: { name: 'Risk-Adjusted Return', value: +rar.toFixed(2), score: Math.round(primaryScore) },
        secondary: [
            { name: 'Margin Events', value: stats.marginEvents || 0, score: Math.round(secondaryScore) }
        ]
    };
}

// ── Generator ──
// Primary: Profit per MW = NetProfit / CapacityMW
// Secondary: Imbalance cost per MWh (lower is better)
function generatorRoleScore(stats, cfg) {
    const profitPerMW = stats.capacityMW > 0 ? stats.netProfit / stats.capacityMW : 0;
    const primaryScore = clamp(mapThreshold(profitPerMW, cfg.breakpoints), 0, 100);

    // Imbalance cost: 0 → 100, reference ceiling → 0
    const imbRef = 50; // £50/MWh reference ceiling
    const imbCostPerMWh = stats.totalMWh > 0 ? Math.abs(stats.imbalanceCost || 0) / stats.totalMWh : 0;
    const secondaryScore = clamp(100 - (imbCostPerMWh / imbRef) * 100, 0, 100);

    const pw = cfg.primaryWeight || 0.80;
    const roleScore = clamp(Math.round(pw * primaryScore + (1 - pw) * secondaryScore), 0, 100);

    return {
        roleScore,
        primary: { name: 'Profit/MW', value: +profitPerMW.toFixed(0), score: Math.round(primaryScore) },
        secondary: [
            { name: 'Imbalance Cost/MWh', value: +imbCostPerMWh.toFixed(1), score: Math.round(secondaryScore) }
        ]
    };
}

// ── BESS ──
// Primary: Revenue per MWh shifted (net of losses/degradation)
// Secondary: BM revenue share + SoC health (no penalties)
function bessRoleScore(stats, cfg) {
    const mwhShifted = Math.max(1, stats.mwhShifted || 1);
    const revPerMWh = stats.netProfit / mwhShifted;
    const primaryScore = clamp(mapThreshold(revPerMWh, cfg.breakpoints), 0, 100);

    // BM share: what % of revenue came from BM (>=50% is excellent)
    const bmShare = stats.totalRevenue > 0 ? (stats.bmRevenue || 0) / stats.totalRevenue : 0;
    const bmShareScore = clamp(bmShare * 200, 0, 100); // 50% → 100

    // SoC health: fewer penalty events = higher
    const socPenalties = stats.socPenalties || 0;
    const socScore = clamp(100 - socPenalties * 20, 0, 100);

    const secondaryScore = (bmShareScore + socScore) / 2;

    const pw = cfg.primaryWeight || 0.75;
    const roleScore = clamp(Math.round(pw * primaryScore + (1 - pw) * secondaryScore), 0, 100);

    return {
        roleScore,
        primary: { name: '£/MWh Shifted', value: +revPerMWh.toFixed(1), score: Math.round(primaryScore) },
        secondary: [
            { name: 'BM Revenue Share', value: +(bmShare * 100).toFixed(0) + '%', score: Math.round(bmShareScore) },
            { name: 'SoC Health', value: socPenalties + ' penalties', score: Math.round(socScore) }
        ]
    };
}

// ── Supplier ──
// Primary: Net cost per MWh after imbalance (lower = better — INVERTED)
// Secondary: Hedge ratio + imbalance as % of cost
function supplierRoleScore(stats, cfg) {
    const costPerMWh = stats.totalMWh > 0 ? Math.abs(stats.netCost || 0) / stats.totalMWh : 80;
    // Inverted breakpoints: lower cost = higher score
    const primaryScore = clamp(mapThreshold(costPerMWh, cfg.breakpoints), 0, 100);

    const hedgeRatio = clamp((stats.hedgeRatio || 0.5) * 100, 0, 100);
    const imbPct = stats.netCost > 0 ? Math.abs(stats.imbalanceCost || 0) / stats.netCost * 100 : 50;
    const imbPctScore = clamp(100 - imbPct * 2, 0, 100); // 0% imbalance → 100, 50%+ → 0

    const secondaryScore = (hedgeRatio + imbPctScore) / 2;

    const pw = cfg.primaryWeight || 0.80;
    const roleScore = clamp(Math.round(pw * primaryScore + (1 - pw) * secondaryScore), 0, 100);

    return {
        roleScore,
        primary: { name: 'Cost/MWh', value: '£' + costPerMWh.toFixed(0), score: Math.round(primaryScore) },
        secondary: [
            { name: 'Hedge Ratio', value: (stats.hedgeRatio * 100 || 0).toFixed(0) + '%', score: Math.round(hedgeRatio) },
            { name: 'Imbalance % of Cost', value: imbPct.toFixed(0) + '%', score: Math.round(imbPctScore) }
        ]
    };
}

// ── DSR Aggregator ──
// Primary: Reliability-adjusted revenue = Revenue × Reliability%
// Secondary: Customer impact penalty score
function dsrRoleScore(stats, cfg) {
    const reliability = clamp(stats.reliability || 1, 0, 1);
    const relAdjRev = (stats.netProfit || 0) * reliability;
    const primaryScore = clamp(mapThreshold(relAdjRev, cfg.breakpoints), 0, 100);

    const missedEvents = stats.missedEvents || 0;
    const secondaryScore = clamp(100 - missedEvents * 15, 0, 100);

    const pw = cfg.primaryWeight || 0.80;
    const roleScore = clamp(Math.round(pw * primaryScore + (1 - pw) * secondaryScore), 0, 100);

    return {
        roleScore,
        primary: { name: 'Reliability-Adj Rev', value: '£' + relAdjRev.toFixed(0), score: Math.round(primaryScore) },
        secondary: [
            { name: 'Missed Events', value: missedEvents, score: Math.round(secondaryScore) }
        ]
    };
}

// ── NESO (Combined System Operator + Market Operator) ──
// Four-part weighted: stability (40%), cost (20%), forecast MAE (15%), market clearing quality (25%)
// Reflects combined role: DA/ID auction clearing + real-time BM dispatch + constraint management
function nesoRoleScore(stats, cfg) {
    // Stability: based on average |NIV| — lower is better (real-time balancing quality)
    const avgAbsNIV = stats.avgAbsNIV || 0;
    const stabilityScore = clamp(100 - (avgAbsNIV / 6.2), 0, 100); // 0 NIV→100, 620→0

    // Cost: lower total system cost is better (market clearing efficiency)
    const costPerSP = stats.totalSystemCost / Math.max(1, stats.totalSPs || 1);
    const costScore = clamp(100 - (costPerSP / 50), 0, 100);

    // Forecast MAE: lower is better
    const maeScore = clamp(100 - (stats.forecastMAE || 0) * 2, 0, 100);

    // Market Clearing Quality: how well DA/ID auctions matched supply to demand
    // Based on: price stability (lower volatility = better), participation rate, bid coverage
    const priceVol = stats.priceVolatility || 0;
    const priceVolScore = clamp(100 - priceVol * 0.5, 0, 100);
    const participation = clamp((stats.participationRate || 0.5) * 100, 0, 100);
    const clearingScore = (priceVolScore + participation) / 2;

    const sw = cfg.stabilityWeight || 0.40;
    const cw = cfg.costWeight || 0.20;
    const mw = cfg.maeWeight || 0.15;
    const clw = cfg.clearingWeight || 0.25;
    const roleScore = clamp(Math.round(sw * stabilityScore + cw * costScore + mw * maeScore + clw * clearingScore), 0, 100);

    return {
        roleScore,
        primary: { name: 'Stability Index', value: Math.round(stabilityScore), score: Math.round(stabilityScore) },
        secondary: [
            { name: 'System Cost/SP', value: '£' + costPerSP.toFixed(0), score: Math.round(costScore) },
            { name: 'Forecast MAE', value: (stats.forecastMAE || 0).toFixed(1), score: Math.round(maeScore) },
            { name: 'Clearing Quality', value: Math.round(clearingScore), score: Math.round(clearingScore) }
        ]
    };
}

// ── Elexon (Settlement Body) ──
// Scores settlement accuracy, timeliness, and transparency
// In simulator context: how accurately the Elexon player identified/reported imbalances
function elexonRoleScore(stats, cfg) {
    // Accuracy: how close settlement calculations match actual metered data
    // Lower settlement error = higher score
    const settlementError = stats.settlementError || 0;
    const accuracyScore = clamp(100 - settlementError * 5, 0, 100);

    // Timeliness: fraction of settlements processed on time
    const onTimeRate = clamp((stats.onTimeRate || 1.0) * 100, 0, 100);

    // Transparency: audit completeness — how many SPs have complete audit trails
    const auditCoverage = clamp((stats.auditCoverage || 1.0) * 100, 0, 100);

    const aw = cfg?.accuracyWeight || 0.50;
    const tw = cfg?.timelinessWeight || 0.30;
    const trw = cfg?.transparencyWeight || 0.20;
    const roleScore = clamp(Math.round(aw * accuracyScore + tw * onTimeRate + trw * auditCoverage), 0, 100);

    return {
        roleScore,
        primary: { name: 'Settlement Accuracy', value: Math.round(accuracyScore) + '%', score: Math.round(accuracyScore) },
        secondary: [
            { name: 'On-Time Rate', value: Math.round(onTimeRate) + '%', score: Math.round(onTimeRate) },
            { name: 'Audit Coverage', value: Math.round(auditCoverage) + '%', score: Math.round(auditCoverage) }
        ]
    };
}

// ── Interconnector ──
// Primary: Congestion revenue
// Secondary: Availability %, stress contribution
function interconnectorRoleScore(stats, cfg) {
    const congestionRev = stats.congestionRevenue || 0;
    const primaryScore = clamp(mapThreshold(congestionRev, cfg.breakpoints), 0, 100);

    const availability = clamp((stats.availability || 1) * 100, 0, 100);
    const stressHelp = clamp((stats.stressContribution || 0) * 20, 0, 100);
    const secondaryScore = (availability + stressHelp) / 2;

    const pw = cfg.primaryWeight || 0.80;
    const roleScore = clamp(Math.round(pw * primaryScore + (1 - pw) * secondaryScore), 0, 100);

    return {
        roleScore,
        primary: { name: 'Congestion Revenue', value: '£' + congestionRev.toFixed(0), score: Math.round(primaryScore) },
        secondary: [
            { name: 'Availability', value: availability.toFixed(0) + '%', score: Math.round(availability) },
            { name: 'Stress Help', value: (stats.stressContribution || 0).toFixed(1), score: Math.round(stressHelp) }
        ]
    };
}

// ─── SYSTEM SCORE ───
// Common for all roles. Measures how much a player helped or hurt system stability.
// Input: { totalNIVContribution, stressWindowHelps, missedDeliveries, causedBlackout }
// Output: 0-100

export function computeSystemScore(metrics) {
    if (!metrics) return 50; // neutral default

    // Base: map NIV contribution percentile
    // Positive contribution = helped (reduced |NIV|), negative = hurt
    const nivBase = clamp(50 + (metrics.totalNIVContribution || 0) * 0.1, 0, 100);

    // Bonus for stress window help
    const stressBonus = Math.min((metrics.stressWindowHelps || 0) * 5, 25);

    // Penalties
    const missedPenalty = (metrics.missedDeliveries || 0) * 10;
    const blackoutPenalty = metrics.causedBlackout ? 40 : 0;

    return clamp(Math.round(nivBase + stressBonus - missedPenalty - blackoutPenalty), 0, 100);
}

// ─── OVERALL SCORE ───
export function computeOverallScore(roleScore, systemScore, alpha) {
    const a = alpha ?? SCORING_CONFIG.alpha ?? 0.6;
    return clamp(Math.round(a * roleScore + (1 - a) * systemScore), 0, 100);
}

// ─── MULTI-ROUND FINAL SCORE ───
// Rewards consistency: mean - penalty * std
export function computeFinalScore(overallScores) {
    if (!overallScores || overallScores.length === 0) return 0;
    const n = overallScores.length;
    const mean = overallScores.reduce((s, v) => s + v, 0) / n;
    if (n < 2) return Math.round(mean);

    const variance = overallScores.reduce((s, v) => s + (v - mean) ** 2, 0) / (n - 1);
    const std = Math.sqrt(variance);
    const penalty = SCORING_CONFIG.consistencyPenalty ?? 0.1;

    return clamp(Math.round(mean - penalty * std), 0, 100);
}
